File: solutions/lua/space-age/1/space-age.lua
local SpaceAge = {}

function SpaceAge:new(seconds)
  local t = {}
  local earth_orbit_seconds = 31557600

  local function get_age(orbital_period_years)
    return tonumber(string.format("%.2f", seconds / (earth_orbit_seconds * orbital_period_years)))
  end

  t.seconds    = seconds
  t.on_mercury = function() return get_age(0.2408467)  end
  t.on_venus   = function() return get_age(0.61519726) end
  t.on_earth   = function() return get_age(1.0)        end
  t.on_mars    = function() return get_age(1.8808158)  end
  t.on_jupiter = function() return get_age(11.862615)  end
  t.on_saturn  = function() return get_age(29.447498)  end
  t.on_uranus  = function() return get_age(84.016846)  end
  t.on_neptune = function() return get_age(164.79132)  end

  self.__index = self
  setmetatable(t, self)
  
  return t
end

return SpaceAge

File: solutions/scala/space-age/2/src/main/scala/SpaceAge.scala
object SpaceAge {

  // Constant names should be in upper camel case, similar to Java’s static final members.
  val EarthYearInSeconds = 31557600

  // Orbital period in Earth years
  val VenusOrbit = 0.61519726
  val MercuryOrbit = 0.2408467
  val MarsOrbit = 1.8808158
  val JupiterOrbit = 11.862615
  val SaturnOrbit = 29.447498
  val UranusOrbit = 84.016846
  val NeptuneOrbit = 164.79132

  def onEarth(ageInSeconds: Double): Double = ageInSeconds / EarthYearInSeconds

  def onVenus(ageInSeconds: Double): Double = onEarth(ageInSeconds) / VenusOrbit

  def onMercury(ageInSeconds: Double): Double = onEarth(ageInSeconds) / MercuryOrbit

  def onMars(ageInSeconds: Double): Double = onEarth(ageInSeconds) / MarsOrbit

  def onJupiter(ageInSeconds: Double): Double = onEarth(ageInSeconds) / JupiterOrbit

  def onSaturn(ageInSeconds: Double): Double = onEarth(ageInSeconds) / SaturnOrbit

  def onUranus(ageInSeconds: Double): Double = onEarth(ageInSeconds) / UranusOrbit

  def onNeptune(ageInSeconds: Double): Double = onEarth(ageInSeconds) / NeptuneOrbit
}
